File: algorithms-java/src/main/java/me/freedom4live/algorithms/primes/ferma/FermaPrimeChecker.java
package me.freedom4live.algorithms.primes.ferma;

import java.math.BigInteger;
import java.util.concurrent.ThreadLocalRandom;

class FermaPrimeChecker {

    static boolean isPrime(long number, long times) {
        if (number < 2) return false;
        if (number == 2) return true;

        BigInteger theNumber = BigInteger.valueOf(number);

        for (int i = 0; i < times; i++) {
            long randomNumber = getRandomNumberForTest(number);

            BigInteger powResult = BigInteger.valueOf(randomNumber).pow((int) number - 1);
            long result = powResult.mod(theNumber).longValue();
            if (result != 1) return false;
        }
        return true;
    }

    private static long getRandomNumberForTest(long number) {
        return ThreadLocalRandom.current().longs(2, number)
                .findFirst().orElseThrow(() -> new IllegalStateException("Can't get current random state!"));
    }

}

File: algorithms-kotlin/src/main/kotlin/me/freedom4live/algorithms/primes/ferma/FermaPrimeChecker.kt
package me.freedom4live.algorithms.primes.ferma

import java.math.BigInteger
import kotlin.random.Random

object FermaPrimeChecker {

    fun isPrime(number: Long, times: Int): Boolean = when {
        number < 2 -> false
        number == 2L -> true
        else -> checkIsPrime(number, times)
    }

    private fun checkIsPrime(number: Long, times: Int): Boolean {
        val bigIntNumber = BigInteger.valueOf(number)

        repeat(times) {
            val randomNumber = getRandom(number).toBigInteger()
            val preResult = randomNumber.pow(number.toInt() - 1)
            val result = preResult.mod(bigIntNumber).toInt()

            if (result != 1) return false
        }

        return true
    }

    private fun getRandom(number: Long): Long = Random.nextLong(2, number)

}
